src/ghci/warning_formatter.rs

```diff
@@ -0,0 +1,236 @@
+//! Warning message formatting with GHC-style coloring.
+//!
+//! This module provides functionality to colorize GHC diagnostic messages
+//! to match the output format and colors used by GHC itself.
+
+use owo_colors::OwoColorize;
+use owo_colors::Stream::Stdout;
+
+use crate::ghci::parse::Severity;
+
+/// Formats warning and error messages with GHC-compatible colors.
+pub struct WarningFormatter;
+
+// TODO: Connect it to the ANSI colors
+// that we need to preserve in https://github.com/MercuryTechnologies/ghciwatch/blob/TrackWarnings/src/ghci/parse/ghc_message/mod.rs#L153
+impl WarningFormatter {
+    /// Create a new warning formatter.
+    pub fn new() -> Self {
+        Self
+    }
+
+    /// Apply GHC-style coloring to a complete diagnostic message.
+    ///
+    /// This processes multi-line messages and applies appropriate coloring
+    /// to each line based on GHC's output patterns.
+    pub fn colorize_message(&self, message: &str, severity: Severity) -> String {
+        let lines: Vec<&str> = message.lines().collect();
+        let mut colored_lines = Vec::new();
+
+        for line in lines {
+            let colored_line = self.colorize_line(line, severity);
+            colored_lines.push(colored_line);
+        }
+
+        colored_lines.join("\n")
+    }
+
+    /// Apply colors to a single line of a diagnostic message based on GHC patterns.
+    ///
+    /// This method recognizes different types of lines in GHC output and applies
+    /// appropriate coloring:
+    /// - Source code lines with line numbers (e.g., "  28 | import Data.Coerce (coerce)")
+    /// - Caret indicator lines (e.g., "     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^")
+    /// - Warning/error flags in brackets (e.g., "[-Wunused-imports]")
+    fn colorize_line(&self, line: &str, severity: Severity) -> String {
+        // Detect different types of lines and apply appropriate coloring
+
+        // Source code lines with line numbers (e.g., "  28 | import Data.Coerce (coerce)")
+        if let Some(pipe_pos) = line.find(" | ") {
```

Comment:
This is kinda hacky, there should at least be a comment here. Ideally we'd integrate this into the parser but I understand that's a more invasive change.

Reply:
Yeah if there was a struct I could plug into with the various pieces of a GHC message, that would be really helpful. We could store original color information there as well. 

I'd like to put that off for a follow up (and left one comment-based pointer for that at the top here). Do you think another comment here would suffice for now?

Reply:
Added a comment here.